application/codeaction/codeaction.go

```diff
@@ -0,0 +1,90 @@
+package codeaction
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/rs/zerolog"
+	"github.com/rs/zerolog/log"
+
+	"github.com/snyk/snyk-ls/application/server/lsp"
+	"github.com/snyk/snyk-ls/domain/ide/converter"
+	"github.com/snyk/snyk-ls/domain/snyk"
+	"github.com/snyk/snyk-ls/internal/uri"
+)
+
+// IssuesProvider is an interface that allows to retrieve issues for a given path and range.
+// This is used instead of any concrete dependency to allow for easier testing and more flexibility in implementation.
+type issuesProvider interface {
+	IssuesFor(path string, r snyk.Range) []snyk.Issue
+}
+
+// CodeActionsService is an application-layer service for handling code actions.
+type CodeActionsService struct {
+	IssuesProvider issuesProvider
+
+	// issuesCache holds all the issues that were returns by the GetCodeActions method.
+	// This is used to resolve the code actions later on in ResolveCodeAction.
+	issuesCache map[uuid.UUID]cachedIssue
```

Comment:
why do we need this?

Reply:
When we get a `codeAction/resolve` message, we need to pull the associated `snyk.CodeAction` to run the deferred edit. Instead of searching again through the folder we can simply cache the issues/code actions between the `textDocument/codeAction` message and the `codeAction/resolve` message.